### PR TITLE
Add validation to frontmatter

### LINKED ISSUE #1374: [Enhancement]: Add validation to frontmatter
### What would you like us to improve?

I am looking to add TypeScript validation to frontmatter.

### How important it is?

This is a quality-of-life improvement. I am looking to contribute and would happily take this on.

### PR DESCRIPTION
**Pull request by simkasss**
This PR depends on [PR #1375 
](https://github.com/smartcontractkit/documentation/pull/1375) 

## Closing issues

closes #1374 
...

## Description

Added validation to frontmatter. Validation happens using [content collections](https://docs.astro.build/en/guides/content-collections/). Creating collection schemas guarantees that specified data exists in a predictable form in frontmatter. 

## Changes

- Collections are created in src/content directory by moving content from src/pages directory
- Collection frontmatter schemas are defined src/content/config.ts
- Frontmatter datatypes are defined with Zod.
- Collections are queried and content is returned content in src/pages directory.

### CODE REVIEW

File: .eslintignore [hunk 1]
```diff
@@ -27,3 +27,4 @@ vercel.json
 tsconfig.json
 
 /src/pages/chainlink-functions/resources/example-source.js
+/src/env.d.ts
```
Comment (DeividasK) [inline .eslintignore:30]: Missing new line at the end of file. Avoid making unnecessary changes.

File: .prettierignore [hunk 2]
```diff
@@ -10,5 +10,5 @@ data-source
 typechain-types
 public/snippets
 public/search-index.json
-src/pages/chainlink-nodes/v1/running-a-chainlink-node.mdx
-vercel.json
\ No newline at end of file
+src/content/chainlink-nodes/v1/running-a-chainlink-node.mdx
```
Comment (DeividasK) [inline .prettierignore:13]: Could you create an issue for this? Arbitrarily adding files to .prettierignore is a bad practice and should be avoided. I'm sure you can solve this rather trivially.

File: src/content/chainlink-automation/register-upkeep.mdx [hunk 3]
```diff
@@ -37,6 +36,7 @@ This guide explains how to register a custom logic Upkeep that uses a [compatibl
      {/* prettier-ignore */}
 
      {" "}
+
```
Comment (DeividasK) [inline src/content/chainlink-automation/register-upkeep.mdx:39]: Remove these new lines. Avoid making unnecessary changes.

File: src/content/config.ts [hunk 4]
```diff
@@ -0,0 +1,214 @@
+import { z, defineCollection } from "astro:content"
+
+// Define a `type` and `schema` for each collection
```
Comment (DeividasK) [inline src/content/config.ts:3]: Redundant?

File: src/content/config.ts [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+            })
+            .optional(),
+        })
+        .optional(),
+      isMdx: z.boolean().optional(),
+    })
+    .strict(),
+})
+
+// Export a single `collections` object to register collections
```
Comment (DeividasK) [inline src/content/config.ts:202]:
```suggestion
```

File: src/layouts/MainLayout.astro [hunk 6]
```diff
@@ -9,7 +9,7 @@ import * as CONFIG from "../config"
 import { NewsletterCTA } from "../components/Footer/NewsletterCTA.tsx"
 import Footer from "~/components/Footer/Footer.astro"
 
-const { content = {}, frontmatter } = Astro.props
+const { frontmatter, frontmatter: content = {} } = Astro.props
```
Comment (DeividasK) [inline src/layouts/MainLayout.astro:12]: Create an issue to remove `content` in favor of using `frontmatter` prop.

File: src/pages/chainlink-nodes.astro [hunk 7]
```diff
@@ -0,0 +1,98 @@
+---
```
Comment (DeividasK) [inline src/pages/chainlink-nodes.astro:1]: Is this change from .mdx to .astro for consistency reasons? I assume this only changes the extension and not the contents of the file.

File: tsconfig.json [hunk 8]
```diff
@@ -26,6 +26,7 @@
       "@components/*": ["src/components/*"],
       "@components": ["src/components/index.ts"],
       "@abi": ["src/features/abi/index.ts"]
-    }
+    },
+    "strictNullChecks": true
```
Comment (DeividasK) [inline tsconfig.json:30]: Why this?

File: src/content/chainlink-automation/register-upkeep.mdx (hunk 3 above)
Comment (simkasss) [reply src/content/chainlink-automation/register-upkeep.mdx:39]: Lines are added/removed by prettier.

File: src/content/chainlink-functions/tutorials/simple-computation.mdx [hunk 9]
```diff
@@ -176,7 +175,7 @@ Read the [Request Configuration](https://github.com/smartcontractkit/functions-h
 
 ### source.js
 
-Read the [JavaScript code](https://github.com/smartcontractkit/functions-hardhat-starter-kit#javascript-code) section for a detailed explanation of how to write compatible JavaScript source code.  The code is self-explanatory and has comments to help you understand all the steps. The main steps are:
+Read the [JavaScript code](https://github.com/smartcontractkit/functions-hardhat-starter-kit#javascript-code) section for a detailed explanation of how to write compatible JavaScript source code. The code is self-explanatory and has comments to help you understand all the steps. The main steps are:
```
[the remark this answers is not in the document]
Comment (simkasss) [reply src/content/chainlink-functions/tutorials/simple-computation.mdx:178]: Redundant space is removed by prettier.

File: src/pages/chainlink-nodes.astro (hunk 7 above)
Comment (simkasss) [reply src/pages/chainlink-nodes.astro:1]: Yes, this change is made for consistency reasons.

File: tsconfig.json (hunk 8 above)
Comment (simkasss) [reply tsconfig.json:30]: By enabling strictNullChecks, you can prevent null-related runtime errors. When strictNullChecks is set to true, the TypeScript compiler enforces stricter rules regarding null and undefined values (null and undefined values are not assignable to other types, etc.)

**Comment by dwightjl** [comment]
We will have https://github.com/smartcontractkit/documentation/pull/1375 merged shortly. I'll need some time to review this PR more thoroughly. Looks like great work. Thank you!

**Comment by simkasss** [comment]
Is there a reason why the links "chainlink-functions/api-reference/FunctionsClient" and "chainlink-functions/api-reference/Functions" are capitalized? These two links fail because Astro generates links in lowercase, but the original links and references to these links are capitalized.

If I change the links and references to these links to lowercase letters, there are no link check errors.
Additionally, I would need to add redirects to the lowercase links (which I haven't done yet). 

Please let me know if I should proceed with adding redirects to lowercase links, or if I should find a way to keep the links capitalized.

**Comment by dwightjl** [comment]
@simkasss Can you rebase this one? We made some significant changes in several docs that have merge conflicts. Now that the CCIP documentation is out, we can get this merged.

**Comment by simkasss** [comment]
The main branch has been merged into this branch, and the 'ccip' Astro collection has been added. 
The capitalized links have been changed to lowercase letters. 
Let me know if I need to make more changes.

**Comment by dwightjl** [comment]
We have several more changes coming in tomorrow morning, so this PR will need to wait another day.

We will review this as a team tomorrow morning after we finish merging some other PRs.

This PR renames some files, so we'll also need to add redirects for those pages.

**Comment by dwightjl** [comment]
> Is there a reason why the links "chainlink-functions/api-reference/FunctionsClient" and "chainlink-functions/api-reference/Functions" are capitalized? These two links fail because Astro generates links in lowercase, but the original links and references to these links are capitalized.

Sorry I just saw the comment.

Lower-case should be fine. Redirects will be necessary, but that's not an issue.

File: src/content/config.ts [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+          title: z.string(),
+          description: z.string(),
+          image: z.object({
+            0: z.string(),
+          }),
+        })
+        .optional(),
+    })
+    .strict(),
+})
```
Comment (fmonper1) [inline src/content/config.ts:24]: Good work on this, I would recommend extracting the base frontmatter to a zod object and  [extend](https://github.com/colinhacks/zod#extend) this base object

```suggestion
const baseFrontmatter = z
    .object({
      section: z.string(),
      date: z.string(),
      title: z.string(),
      isMdx: z.boolean(),
      permalink: z.string(),
      whatsnext: z.record(z.string(), z.string()).optional()
      })

const anyApiCollection = defineCollection({
  type: "content",
  schema: baseFrontmatter.extend({
      metadata: z
        .object({
          title: z.string(),
          description: z.string(),
          image: z.object({
            0: z.string(),
          }),
        })
        .optional(),
    })
    .strict(),
})
```

File: src/content/config.ts [hunk 11]
```diff
@@ -0,0 +1,225 @@
+import { z, defineCollection } from "astro:content"
+
+const anyApiCollection = defineCollection({
+  type: "content",
+  schema: z
+    .object({
+      section: z.string(),
```
Comment (fmonper1) [inline src/content/config.ts:7]: There is only a limited amount of sections that can be used throughout the app, so maybe we could narrow down this type to a [zod enum](https://github.com/colinhacks/zod#zod-enums) for extra safety

File: src/config/sidebar.ts
Comment (fmonper1) [inline src/config/sidebar.ts:1]: If we create a zod enum for the section frontmatter, we can type this sidebar to something like:

```
const sectionEnum = z.enum(["gettingStarted", "ccip"])

type Sections = z.infer<typeof sectionEnum>; // "gettingStarted" | "ccip"

type SectionEntry = {
  section: string
  contents: {title: string, url: string}[]
}

const SIDEBAR: Record<Sections, SectionEntry[]> = ... 
```

File: src/content/config.ts (hunk 10 above)
Comment (simkasss) [reply src/content/config.ts]: I have simplified this even further and created BaseFrontmatter and BaseCollection. All the content collections are inferred from BaseCollection. 
However, it would be beneficial to determine which frontmatter parameters are required and which ones are optional. I have declared the parameters that are not present in all files as optional. 
@fmonper1, could you take a look and share your opinion?

**Comment by dwightjl** [comment]
Looks like the TOC isn't being generated: https://documentation-git-fork-simkasss-add-types-chainlinklabs.vercel.app/vrf/v2/best-practices

The Data Feeds TOC renders properly because it gets applied after the `<h2>` tags appear on the feeds list: https://docs.chain.link/data-feeds/price-feeds/addresses#Sepolia%20Testnet

File: src/content/config.ts [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+])
+
+const baseFrontmatter = z
+  .object({
+    section: sectionEnum,
+    date: z.string().optional(),
+    title: z.string(),
+    whatsnext: z.record(z.string(), z.string()).optional(),
+    isMdx: z.boolean().optional(),
+    permalink: z.string().optional(),
```
Comment (fmonper1) [inline src/content/config.ts:24]: Permalink should not be needed as this is legacy from 11ty

**Comment by dwightjl** [comment]
@simkasss everything appears to be working well. Are there any remaining changes required?

**Comment by simkasss** [comment]
@dwightjl , I think you can merge it.

**Comment by dwightjl** [comment]
Looks like this change has issues with our link checker improvements: https://github.com/smartcontractkit/documentation/pull/1420

I'll see if I can find out what's causing this.

**Review by dwightjl — CHANGES_REQUESTED** [review]
Looks like we might need to add redirects for the new paths.

**Comment by dwightjl** [comment]
Rebased and force-pushed to resolve remaining conflicts.

**Comment by dwightjl** [comment]
@simkasss Do we want to include .astro/ in .gitignore? Or should we be pushing that to source?

**Comment by dwightjl** [comment]
I think this looks ready. I just need to re-check everything on staging again. @fmonper1 if you have some time can you also take a look?

File: src/layouts/MainLayout.astro [hunk 13]
```diff
@@ -9,7 +9,7 @@ import * as CONFIG from "../config"
 import { NewsletterCTA } from "../components/Footer/NewsletterCTA"
 import Footer from "~/components/Footer/Footer.astro"
 
-const { content = {}, frontmatter } = Astro.props
+const { frontmatter, frontmatter: content = {} } = Astro.props
```
Comment (fmonper1) [inline src/layouts/MainLayout.astro:12]: Question:

Why are we destructuring twice the same property? Is this intended or is this a typo?

**Review by fmonper1 — COMMENTED** [review]
LGTM, there is just one comment regarding destructuring

File: src/layouts/MainLayout.astro (hunk 13 above)
Comment (simkasss) [reply src/layouts/MainLayout.astro]: I am not sure why, but before this PR, it seems that the `content` object was used as a `frontmatter` in the MainLayout. So after I moved all the pages to src/content, I had to assign the `content` to the frontmatter value, so that the MainLayout would work as intended. 
I have modified the MainLayout by removing the 'content' variable and replacing it with `frontmatter` instead. Seems like it works without any issues as before 🤷‍♀️ 

However, I am not sure why the `content` object was used in this case. I tried to search in Astro Docs, but I couldn't find the answer.

**Comment by dwightjl** [comment]
I'm doing a final check on everything now.

**Review by dwightjl — APPROVED** [review]
